src/pycell.rs

```diff
@@ -1,6 +1,144 @@
+//! Rust has strict aliasing rules - you can either have any number of immutable (shared) references or one mutable
+//! reference. Python's ownership model is the complete opposite of that - any Python object
+//! can be referenced any number of times, and mutation is allowed from any reference.
+//!
+//! PyO3 deals with these differences by employing the [Interior Mutability]
+//! pattern. This requires that PyO3 enforces the borrowing rules and it has two mechanisms for
+//! doing so:
+//! - Statically it can enforce threadsafe access with the [`Python<'py>`](crate::Python) token.
+//! All Rust code holding that token, or anything derived from it, can assume that they have
+//! safe access to the Python interpreter's state. For this reason all the native Python objects
+//! can be mutated through shared references.
+//! - However, methods and functions in Rust usually *do* need `&mut` references. While PyO3 can
+//! use the [`Python<'py>`](crate::Python) token to guarantee thread-safe access to them, it cannot
+//! statically guarantee uniqueness of `&mut` references. As such those references have to be tracked
+//! dynamically at runtime, using [`PyCell`] and the other types defined in this module. This works
+//! similar to std's [`RefCell`](std::cell::RefCell) type.
+//!
+//! # When *not* to use PyCell
+//!
+//! Usually you can use `&mut` references as method and function receivers and arguments, and you
+//! won't need to use [`PyCell`] directly:
+//!
+//! ```rust
+//! use pyo3::prelude::*;
+//!
+//! #[pyclass]
+//! struct Number {
+//!     inner: u32,
+//! }
+//!
+//! #[pymethods]
+//! impl Number {
+//!     fn increment(&mut self) {
+//!         self.inner += 1;
+//!     }
+//! }
+//! ```
+//!
+//! The [`#[pymethods]`](crate::pymethods) proc macro will generate this wrapper function (and more),
+//! using [`PyCell`] under the hood:
+//!
+//! ```ignore
+//! // This function is exported to Python.
+//! unsafe extern "C" fn __wrap(slf: *mut PyObject, _args: *mut PyObject) -> *mut PyObject {
+//!     pyo3::callback::handle_panic(|py| {
+//!         let cell: &PyCell<Number> = py.from_borrowed_ptr(slf);
+//!         let mut _ref: PyRefMut<Number> = cell.try_borrow_mut()?;
+//!         let slf: &mut Number = &mut _ref;
+//!         pyo3::callback::convert(py, Number::increment(slf))
+//!     })
+//! }
+//!  ```
+//!
+//! # When to use PyCell
+//! ## Using pyclasses from Rust
+//!
+//! However, we *do* need [`PyCell`] if we want to call its methods from Rust:
+//! ```rust
+//! # use pyo3::prelude::*;
+//! #
+//! # #[pyclass]
+//! # struct Number {
+//! #     inner: u32,
+//! # }
+//! #
+//! # #[pymethods]
+//! # impl Number {
+//! #     fn increment(&mut self) {
+//! #         self.inner += 1;
+//! #     }
+//! # }
+//! # fn main() -> PyResult<()> {
+//! Python::with_gil(|py| {
+//!     let n = PyCell::new(py, Number { inner: 0 })?;
```

Comment:
Can we start from a `Py<Number>` or instead? That is much more common that constructing a `PyCell` ourselves.

Reply:
Done :)